### ISSUE TITLE
edd_has_user_purchased should have a second param for bundles

### ISSUE THREAD

**Issue by chriscct7** [easydigitaldownloads/easy-digital-downloads#3786]
edd_has_user_purchased should have a second param, default false to include purchases via a bundle

**Comment by pippinsplugins** [comment]
What's a scenario someone wouldn't want bundles included?

**Comment by chriscct7** [comment]
I can't think of one immediately that's a super good reason tbh. Including bundles makes the function less performant, but I was mostly suggesting that because of backwards compatiblity (we don't include bundles atm). I don't have a reason not for it to support bundles out of the box without an extra param

**Comment by pippinsplugins** [comment]
Ok :+1:

**Comment by sunnyratilal** [comment]
`edd_has_user_purchased` currently works perfectly fine for bundles.

@chriscct7 are you suggesting to have a param that _excludes_ bundles?

**Comment by chriscct7** [comment]
At the time this issue was created bundles were not supported because the function used edd_get_payment_meta_downloads which at the time parsed the meta key. Once EDD Payment was added about 3 months after this ticket was made, the edd_get_payment_meta_downloads function was updated to use EDD Payment and edd_has_user_purchased itself was changed to use EDD Payment in https://github.com/easydigitaldownloads/easy-digital-downloads/commit/f3589e0a49c7f74d763d5aa40a0e04a200d616e9#diff-0918bf07a5a79591afe7faf227032c8fR195. Prior to this change, the function didn't include downloads in a bundle product, because that information wasn't directly stored in the meta key (you had to take the download id of the bundle and then query for the downloads in the bundles), which is why the parameter to exclude this was suggested (as opposed to just adding support but not offering a way to turn this off).

As bundles are now fully supported, this ticket is now invalid. Closing as such.

**Comment by chriscct7** [comment]
IMO, we should use the invalid tag only when the original ticket was invalid, that way we can separate tickets we get often that are invalid versus tickets that were valid at some point but are no longer valid.

**Comment by stephywells** [comment]
This doesn't appear to be working. If I check edd_has_user_purchased with the id of a download in the bundle, it always returns false. In this case, it is comparing the id of the bundle rather than any downloads linked to the bundle.

$purchased_files = $payment->cart_details;
foreach ( $purchased_files as $download ) {
// $download['id'] = the bundle id
if ( in_array( $download['id'], $downloads ) ) {

Is cart_details supposed to include all the ids of all the downloads in the bundle? If so, what would happen when the settings change? It looks like edd_get_bundled_products() is still required there in order to get the downloads linked to the bundle.

**Comment by cklosowski** [comment]
@stephywells No, the `cart_details` should not have all the IDs of the items in the bundle. That's left for retrieving later so that you can change the bundle without having to rerun all the purchases of it to recalculate the data.

Going to re-open this for further investigation and fixing.

**Comment by pippinsplugins** [comment]
We need a better way of determining if a customer has access to an item that is included inside of a bundle, though I'm not certain `edd_has_user_purchased()` is the proper place for it.

Is having purchased a bundle the same as having purchased an item inside of a bundle? In terms of file access, yes, but in other ways not so much.

Is it worth separating them?
